Improve login redirects and step tracking in stepper

daledupreez opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to:
* https://github.com/Automattic/wp-calypso/pull/91074
* https://github.com/Automattic/wp-calypso/pull/90711

Fixes https://github.com/Automattic/wp-calypso/issues/91243

## Proposed Changes

* This PR introduces some key concepts into the core Stepper framework to allow us to better handle logged out users and the redirects that stem from them. The key changes are as follows:
  - Stepper steps now have an optional `requiresLoggedInUser` flag that should be specified in `flow.useSteps()`
    * The PR includes `stepWithRequiredLogin()` and `stepsWithRequiredLogin()` helpers to set this flag easily
  - The pre-existing `StepRoute` component now checks whether the current step has a truthy `requiresLoggedInUser` value, and will not render any content if the user is also not logged in, as well as triggering a redirect to log the user in.
    * The login redirect relies on a new `useLoginUrlForFlow()` hook that builds on my explorations in https://github.com/Automattic/wp-calypso/pull/90711
  - The logic to trigger `calypso_signup_step_start` Tracks events has been moved from `FlowRenderer` to `StepRoute`, and we won't trigger the Tracks event if the `StepRoute` logic from above performs a login redirect for a step that requires a login

## Why are these changes being made?

<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* We've been having internal discussions around improvements to the way Stepper handles Tracks and logins, and one of the key items that was flagged is that the current implementation can send `calypso_signup_step_start` Tracks events _even when the user doesn't see the step_. This PR attempts to resolve this by "knowing" whether a specific step requires the user to be logged in, and performing a redirect when necessary, as well as skipping rendering and triggering a Tracks event when we know a redirect should occur.
  - For now, the new skip logic only occurs when the step is "opted in" to the new requirement, but it should be relatively easy to roll out the behaviour on a flow-by-flow basis.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run this branch locally or via Calypso.live
* Open a window where you are not logged in
* Ensure you have some way of seeing Tracks events
* Navigate to any of the URLs at or below `/setup/site-migration`, including:
  - `/setup/site-migration/`
  - `/setup/site-migration/site-migration-identify`
  - `/setup/site-migration/site-migration-import-or-migrate`
  - `/setup/site-migration/site-migration-upgrade-plan`
  - `/setup/site-migration/site-migration-instructions-i2`
* Verify that you're redirected to the login page, and the redirect URL will return you to your original location. Also verify that no `calypso_signup_step_start` Tracks event was recorded for the `site-migration` flow.
* Add URL parameters to the URLs above, and verify that you are still redirected to login and these URL parameters are included in the redirect URL used by the login page. You should still see no `calypso_signup_step_start` Tracks events for the `site-migration` flow.
* Specify a valid `locale` URL parameter for some of the above URLs, and verify that you're redirected to the right login page with the locale in the path _and_ the redirect URL for the login page includes the relevant non-English locale.  You should still see no `calypso_signup_step_start` Tracks events for the `site-migration` flow.
* In a second window where you are logged in, work through the same set of URLs above and verify that you're seeing the `calypso_signup_step_start` event being fired correctly for the `site-migration` flow
* In the same logged-in window, work through some other flows and verify that you're seeing the `calypso_signup_step_start` event being fired correctly

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

daledupreez committed:
Initial exploration: step Tracks and login checks in StepRoute

daledupreez committed:
Add helpers to add requiresLoggedInUser: true

gabrielcaires committed:
Use hooks to simplify the test setup

gabrielcaires committed:
Add tests to step-route

gabrielcaires committed:
Add temporary rerenter step test

gabrielcaires committed:
Add tests to cover re-entering case

gabrielcaires committed:
Skip tracking when the step content is not rendered

gabrielcaires commented:
@daledupreez I started adding tests to your change and found a scenario I want to confirm.

Should we skip the tracking when the renderStep returns a null content?  

[433160a](https://github.com/Automattic/wp-calypso/pull/91241/commits/433160adcfec924765804728b655534cc27374b3)

gabrielcaires committed:
Add temporary todo with improvement ideas

gabrielcaires commented:
@daledupreez  I committed this temporary change [355cce9](https://github.com/Automattic/wp-calypso/pull/91241/commits/355cce960bf9d46ecb31f77a438034d0bd0a26b4) with some ideas to clean up the code based on this change.

alshakero commented on client/landing/stepper/declarative-flow/helpers/get-step-old-slug.ts:
```diff
@@ -0,0 +1,10 @@
+/**
+ * This can be used when renaming a step. Simply add a map entry with the new step slug and the old step slug and Stepper will fire `calypso_signup_step_start` events for both slugs. This ensures that funnels with the old slug will still work.
+ */
+export const getStepOldSlug = ( stepSlug: string ): string | undefined => {
+	const stepSlugMap: Record< string, string > = {
+		'create-site': 'site-creation-step',
+	};
+
+	return stepSlugMap[ stepSlug ];
+};
```
Maybe just make this a map?
```suggestion
export const oldStepSlugs = {
	'create-site': 'site-creation-step',
} as const;
```

daledupreez replied:
I just moved this for the PR -- I think we should refactor in a follow-up, as this is called from at least two places.

alshakero commented on client/landing/stepper/declarative-flow/internals/components/step-route/index.tsx:
```diff
@@ -11,9 +28,92 @@ type StepRouteProps = {
+			return;
+		}
+
+		// If we're redirecting in the effect above, skip any step or page tracking.
+		if ( step.requiresLoggedInUser && ! userIsLoggedIn ) {
+			return;
+		}
```
Beautiful ❤️ 

Maybe encode this in a variable for readability.

```js
const shouldRedirectToLogin = step.requiresLoggedInUser && ! userIsLoggedIn
```
Then use it in all the 3 places.

daledupreez replied:
Good idea! 💡

gabrielcaires replied:
Idea implemented!

alshakero commented on client/landing/stepper/declarative-flow/internals/components/step-route/index.tsx:
```diff
@@ -11,9 +28,92 @@ type StepRouteProps = {
+		}
+
+		// Also record page view for data and analytics
+		const pathname = window.location.pathname || '';
```
Can `pathname` ever be falsy?

daledupreez replied:
You're right -- this is probably unnecessary -- I think `location.pathname` will always be a string.

gabrielcaires replied:
Removed!

alshakero commented on client/landing/stepper/declarative-flow/internals/components/step-route/index.tsx:
```diff
@@ -11,9 +28,92 @@ type StepRouteProps = {
+//TODO: Check we can move RenderStep function to here and remove the renderStep prop
 const StepRoute = ( { step, flow, showWooLogo, renderStep }: StepRouteProps ) => {
+	const userIsLoggedIn = useSelector( isUserLoggedIn );
+	const loginUrl = useLoginUrlForFlow( { flow } );
```
I think the login URL should bring the user back to the exact step, not only to the flow. Because this now supports logging in at the nth step (not strictly the first).

daledupreez replied:
I had this pushing in the step to start with, but I then changed the implementation inside `useLoginUrlForFlow()` to use whatever the current location is (as returned from `useLocation()`), rather than trying to derive or reconstruct the URL. I'd be happy to change it back, but it felt likely to be more solid/reliable if we started from the current URL. (It may be better to use `window.location.pathname + window.location.search` as a starting point, but the result should be similar.)

alshakero replied:
Oh that's certainly less maintenance.

alshakero commented on client/landing/stepper/declarative-flow/internals/components/step-route/test/index.tsx:
```diff
@@ -0,0 +1,175 @@
+
+			await waitFor( () =>
+				expect( window.location.assign ).toHaveBeenCalledWith(
+					'/start/account/user-social?variationName=some-flow&redirect_to=%2F&toStepper=true'
```
We should probably update this to include the step name.

daledupreez replied:
This feels like it's an artifact of the way the rendering is done. Might it make sense to parameterise either `basename` or `initialEntries` in the `MemoryRouter` so we can clearly control/dictate what the current location will show up as?

Or should we push that logic into separate tests for `useLoginUrlForFlow()`?

alshakero replied:
Nah, this comment is irrelevant now that you clarified that we're just using the current URL to bring the user back to.

alshakero commented on client/landing/stepper/declarative-flow/internals/components/step-route/index.tsx:
```diff
@@ -11,9 +28,92 @@ type StepRouteProps = {
+		( state ) => site && isRequestingSite( state, siteSlugOrId )
+	);
+	const intent = useIntent();
+	const design = useSelectedDesign();
```
Out of scope but I really dislike Stepper being aware of designs. Tracks should probably support a middleware to allow deeper code to intercept events, and add extra information.

daledupreez replied:
💯 Fully agreed! But I didn't want to break the Tracks events as part of trying to fix them! 🙃

alshakero commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -96,15 +71,8 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
-		? !! selectedSite && ! isRequestingSelectedSite
-		: true;
 
+	// TODO: Check if it we still need to check the selected site to preload the steps.
```
Yes, we do. This is just a [measure](https://github.com/Automattic/wp-calypso/pull/91008) to defer preloading steps until the more important loading of the site is done. We never really needed it.

gabrielcaires replied:
Thanks for the context! No changes applied to this logic!

alshakero commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -182,46 +150,11 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
 		}
 	}, [ flow, ref, isFlowStart ] );
 
-	useEffect( () => {
-		// We record the event only when the step is not empty. Additionally, we should not fire this event whenever the intent is changed
-		if ( ! currentStepRoute || ! hasRequestedSelectedSite ) {
-			return;
-		}
-
-		const signupCompleteFlowName = getSignupCompleteFlowNameAndClear();
-		const signupCompleteStepName = getSignupCompleteStepNameAndClear();
-		const isReEnteringStep =
-			signupCompleteFlowName === flow.name && signupCompleteStepName === currentStepRoute;
-		if ( ! isReEnteringStep ) {
-			recordStepStart( flow.name, kebabCase( currentStepRoute ), {
-				intent,
-				is_in_hosting_flow: isAnyHostingFlow( flow.name ),
-				...( design && { assembler_source: getAssemblerSource( design ) } ),
-			} );
-
-			if ( stepOldSlug ) {
-				recordStepStart( flow.name, kebabCase( stepOldSlug ), {
-					intent,
-					is_in_hosting_flow: isAnyHostingFlow( flow.name ),
-					...( design && { assembler_source: getAssemblerSource( design ) } ),
-				} );
-			}
-		}
-
-		// Also record page view for data and analytics
-		const pathname = window.location.pathname || '';
-		const pageTitle = `Setup > ${ flow.name } > ${ currentStepRoute }`;
-		recordPageView( pathname, pageTitle );
-
-		// We leave out intent from the dependency list, due to the ONBOARD_STORE being reset in the exit flow.
-		// This causes the intent to become empty, and thus this event being fired again.
-		// eslint-disable-next-line react-hooks/exhaustive-deps
-	}, [ flow.name, currentStepRoute, hasRequestedSelectedSite ] );
```
❤️

alshakero commented on client/landing/stepper/declarative-flow/internals/types.ts:
```diff
@@ -46,6 +46,10 @@ export type DeprecatedStepperStep = {
 	 * The step slug is what appears as part of the pathname. Eg the intro in /setup/link-in-bio/intro
 	 */
 	slug: string;
+	/**
+	 * Does the step require a logged-in user?
+	 */
+	requiresLoggedInUser?: boolean;
```
We should make a new type entirely that requires (`requiresLoggedInUser`) and deprecate the existing ones, so people get deprecation warnings from TS and the flows would organically upgrade to the new type.

I do something like that [here](https://github.com/Automattic/wp-calypso/blob/3a94454b152d4a683c7c69582d212f59d0d91488/client/landing/stepper/declarative-flow/internals/types.ts#L70).

daledupreez replied:
I intentionally left this as optional for now, mostly because the login requirement is actually flow-specific. So I don't think we should define the login requirement as part of the core step definition. (More on that below.)

alshakero replied:
You're right. Maybe it should be `requiresLoggedInUser?: true`?

alshakero commented on client/landing/stepper/hooks/use-login-url-for-flow.ts:
```diff
@@ -0,0 +1,30 @@
+	const location = useLocation();
+	const { siteId, siteSlug } = useSiteData();
+
+	const redirectTo = addQueryArgs( location.pathname, {
+		...( locale && locale !== 'en' ? { locale } : {} ),
+		...( siteId ? { siteId } : {} ),
+		...( siteSlug ? { siteSlug } : {} ),
+		...Object.fromEntries( new URLSearchParams( location.search ).entries() ),
+	} );
```
100% a nit, but I would spread this into 4 `ifs` 
```suggestion
       let redirectTo = location.pathname;
       
       if ( locale && locale !== 'en' ) {
           redirectTo = addQueryArgs( location.href, { locale } );
       }
       if ( siteId ) {
           redirectTo = addQueryArgs( location.href, { siteId } );
       }
       if ( siteSlug ) {
       	    redirectTo = addQueryArgs( location.href, { siteSlug } );
       }
       if ( siteSlug ) {
       	    redirectTo = addQueryArgs( location.href, { siteSlug } );
       }    
```

And I think if we use `location.href` (vs `pathname`) we don't need to consider the pre-existing params manually.

alshakero commented on client/landing/stepper/declarative-flow/site-migration-flow.ts:
```diff
@@ -41,7 +42,7 @@ const siteMigration: Flow = {
 			: [];
 
-		return [ ...baseSteps, ...hostedVariantSteps ];
+		return stepsWithRequiredLogin( [ ...baseSteps, ...hostedVariantSteps ] );
```
I think the `stepsWithRequiredLogin` utility slightly damages the simple understanding of which step require auth and which don't. Or the one-to-one connection between a given step and auth. It looks as if it will hide all of these steps behind an auth-wall. 

And since these steps are in the `STEPS` constant, it would be better to fix them centrally and set an example for new steps IMHO.

daledupreez replied:
I think a big problem here is that a step on its own can't always dictate whether a user must be logged in or not. As an example, we have onboarding flows that require a user to be logged in for all steps (which is the case here), but some of these steps are used in signup flows where we may want to defer the login/account/user step.

I don't think the `STEPS` constant is the right place to make these determinations for all steps. For some steps, it's definitely required that the user be logged in, regardless of context, so it may make sense for those steps to have the new flag by default.

daledupreez replied:
Also, most flows expect all steps to require authentication, so I wanted a clear shorthand for making that happen at a flow level, especially as we roll this out. I'd prefer to ship the initial change to a small set of flows, and then roll it out to more flows once we feel more confident about it in production.

I think we need the shorthand, but I don't have a strong opinion as to what it should look like -- this came up because I didn't want to really mess up the clean array of `STEPS.*` references in `flow.useSteps()`!

alshakero replied:
Oh, yeah! I didn't consider the fact that the same step may or may not require login based on the flow. Let's keep it as is.

alshakero reviewed (dismissed):
I can't tell you how much I appreciate this. Thank you so so much sir. Left a few comments here and there.

daledupreez commented on client/landing/stepper/declarative-flow/internals/components/step-route/index.tsx:
```diff
@@ -11,9 +28,92 @@ type StepRouteProps = {
+	// Ensure that the selected site is fetched, if available. This is used for event tracking purposes.
+	// See https://github.com/Automattic/wp-calypso/pull/82981.
+
+	// TODO: Check if it is duplicated with the useSiteData hook
```
Good call -- we should be able to get the `Site` object from `useSiteData()`.

gabrielcaires replied:
I confirmed is possible to use the same site reference instead of reading again from the store. 👍

daledupreez commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -182,46 +150,11 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
 		state: AssertConditionState.SUCCESS,
 	};
 
+	//TODO: Move it to step route
```
Maybe? Will it be messier to pass in this function or to pass through `assertCondition`, `stepComponents`/`StepComponent`, `stepData`, and `stepNavigation`?

gabrielcaires replied:
We can improve it on the PR, since it is not mandatory to this one.

daledupreez commented on client/landing/stepper/declarative-flow/internals/components/step-route/index.tsx:
```diff
@@ -11,9 +28,92 @@ type StepRouteProps = {
+			return;
+		}
+
+		if ( ! stepContent ) {
+			return;
+		}
```
I think this is a good addition, though it should be extremely unlikely in practice.
We should also add `stepContent` to the deps.

gabrielcaires replied:
It will prevent us from duplicating the tracking on our  13 flows returning AssertConditionState.FAILURE

daledupreez commented on client/landing/stepper/declarative-flow/internals/components/step-route/test/index.tsx:
```diff
@@ -0,0 +1,175 @@
+			expect( recordPageView ).not.toHaveBeenCalled();
+		} );
+
+		it.todo( 'skips when there is no site slug or ID', () => {} );
```
This syntax is making Jest unhappy - we should clean this up.

gabrielcaires replied:
Removed!

daledupreez commented on client/landing/stepper/declarative-flow/site-migration-flow.ts:
```diff
@@ -41,7 +42,7 @@ const siteMigration: Flow = {
+		return stepsWithRequiredLogin( [ ...baseSteps, ...hostedVariantSteps ] );
 	},
 
 	useAssertConditions(): AssertConditionResult {
```
We should remove the logged in checks and redirects from `useAssertConditions()` so we're more confident that our new handling is doing what it should.

alshakero commented on client/landing/stepper/declarative-flow/internals/components/step-route/test/index.tsx:
```diff
@@ -0,0 +1,175 @@
+			expect( recordPageView ).not.toHaveBeenCalled();
+		} );
+
+		it.todo( 'skips when there is no site slug or ID', () => {} );
```
Apparently this fixes it. 

```suggestion
		it.todo( 'skips when there is no site slug or ID' );
```

gabrielcaires replied:
It will not be part of the final PR

gabrielcaires committed:
Fix lack of "/setup" prefix when code is running on the browser.

Due to some weird issue, when the location.pathname runs on the browser context, it doesnt return the react router base path, in our case /setup.
The best way if found to fix it is using the react-router useHref to generate the path.

I added some tests to cover this case because is very easy to break a lot of stuff and very hard to debug.

gabrielcaires committed:
Move tracking to a separated hook

gabrielcaires committed:
Remove log in logic from site-migration flow

gabrielcaires committed:
Fix typescript issue

gabrielcaires commented:
@daledupreez  Updates:
* I only found 1 bug fixed on this commit [this commit ](https://github.com/Automattic/wp-calypso/pull/91241/commits/eb2f4c0d1c2e8cccab238e8e702a6e115eaa97b4).

* I tested the following scenarios, fixed some minor feedback, and reorganized the code a bit. 

* I also already removed the authorization logic from the site-migration flow here [944e22b](https://github.com/Automattic/wp-calypso/pull/91241/commits/944e22b24357b4bbe8732327731f45b9439f8884) 
* I left other possible improvements like the renderStep to another PR. 


| Scenario | Status | 
|--------|--------|
| Logged In user accessing a protected step | ✅  Step rendered with success + tracking step_start |
| Logged In user accessing a non-protected step | ✅ Step rendered with success  + tracking step_start |
| Anonymous user access a protected step | ✅  Redirected to the login page and the step tracking is not triggered  |
| Anonymous user access a non-protected step | ✅ Step rendered with success  + tracking step_start | 
| Anonymous use authenticating using oath | ✅ Redirect + Step rendered with success  + tracking step_start, after the login | 
| Anonymous use authenticating email | ✅ Redirect + Step rendered with success  + tracking step_start, after the login |

gabrielcaires committed:
Fix typescript issue

daledupreez commented on client/landing/stepper/declarative-flow/internals/components/step-route/test/index.tsx:
```diff
@@ -0,0 +1,182 @@
+import React, { FC, Suspense } from 'react';
+import { MemoryRouter } from 'react-router';
+import recordStepStart from 'calypso/landing/stepper/declarative-flow/internals/analytics/record-step-start';
+import {
+	Flow,
+	StepperStep,
+	StepProps,
+} from 'calypso/landing/stepper/declarative-flow/internals/types';
```
Nit: these are all type imports.
```suggestion
import type {
	Flow,
	StepperStep,
	StepProps,
} from 'calypso/landing/stepper/declarative-flow/internals/types';
```

gabrielcaires replied:
All missing 'type' fixed here 5957f4e698ebdca226b315732e85a21c72da1b06

daledupreez commented on client/landing/stepper/declarative-flow/internals/components/step-route/test/index.tsx:
```diff
@@ -0,0 +1,182 @@
+import { isUserLoggedIn } from 'calypso/state/current-user/selectors';
+import { renderWithProvider } from 'calypso/test-helpers/testing-library';
+import StepRoute from '../';
+import { NavigationControls } from '../../../types';
```
Nit: This is another type.
```suggestion
import type { NavigationControls } from '../../../types';
```

gabrielcaires replied:
All missing 'type' fixed here 5957f4e698ebdca226b315732e85a21c72da1b06

gabrielcaires committed:
Improve tests

daledupreez commented on client/landing/stepper/declarative-flow/internals/components/step-route/index.tsx:
```diff
@@ -11,9 +16,31 @@ type StepRouteProps = {
+		return null;
+	}
+
+	//TODO: Extract the layout logics to another component;
```
I am not sure what you mean by this comment. Could you be more explicit about which layout logic should be moved/refactored?

gabrielcaires replied:
Comment removed [e576a62](https://github.com/Automattic/wp-calypso/pull/91241/commits/e576a62fabfcf7292ce645b26f784a6a4c19c3c9)

daledupreez commented on client/landing/stepper/declarative-flow/internals/components/step-route/test/index.tsx:
```diff
@@ -0,0 +1,193 @@
+jest.mock( 'calypso/landing/stepper/hooks/use-selected-design' );
+jest.mock( 'calypso/lib/analytics/page-view' );
+
+const RegularStep: StepperStep = {
+	slug: 'some-step-slug',
+	asyncComponent: jest.fn(), // For the purpose of this test, we don't need to render the using lazy strategy
+};
+
+const RequiresLoginStep: StepperStep = {
+	...RegularStep,
+	requiresLoggedInUser: true,
+};
+
+const FakeStep: FC< StepProps > = () => {
+	return <div>Step Content</div>;
+};
+
+const FakeFlow: Flow = {
+	name: 'some-flow',
+	useSteps: () => [ RegularStep ],
+	useStepNavigation: () => ( {
+		submit: () => {},
+	} ),
+
+	isSignupFlow: false,
+};
```
Minor note: can we use either `ALL_CAPS` names or a lower-case first letter for the non-component constants? At a glance, it's hard to determine that `RegularStep`, `RequiresLoginStep`, and `FakeFlow` are constants, but `FakeStep` is a component.

gabrielcaires replied:
Fixed here d3aec1b10bc32f9b0ed87a8f4f6ff6c99c7abdac

daledupreez commented on client/landing/stepper/declarative-flow/internals/components/step-route/test/index.tsx:
```diff
@@ -0,0 +1,193 @@
+		expect( getByText( 'Step Content' ) ).toBeInTheDocument();
+	} );
+
+	it( 'doesnt the step content when render step returns null', async () => {
```
Nit: we're missing a `render`.
```suggestion
	it( 'does not render the step content when renderStep() returns null', async () => {
```

daledupreez commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -96,15 +71,8 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
-		? !! selectedSite && ! isRequestingSelectedSite
-		: true;
 
+	// TODO: Check if it we still need to check the selected site to preload the steps.
```
Can we remove the TODO?
```suggestion
```

daledupreez commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -182,46 +150,11 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
 		state: AssertConditionState.SUCCESS,
 	};
 
+	//TODO: Move it to step route
```
To be 100% explicit, can we update the comment to mention the specific component?
```suggestion
	// TODO: Move this logic into the StepRoute component.
```

gabrielcaires replied:
Comment remove here [e576a62](https://github.com/Automattic/wp-calypso/pull/91241/commits/e576a62fabfcf7292ce645b26f784a6a4c19c3c9) since I already opened a new PR addressing it. https://github.com/Automattic/wp-calypso/pull/91345

daledupreez reviewed (commented):
I have some feedback from inspection. I'll start some testing shortly.

gabrielcaires commented:
@daledupreez  I created a second PR, based on this on just to address the renderStep issue. https://github.com/Automattic/wp-calypso/pull/91345

gabrielcaires committed:
Update client/landing/stepper/declarative-flow/internals/index.tsx

Co-authored-by: daledupreez <dale@automattic.com>

gabrielcaires committed:
Update client/landing/stepper/declarative-flow/internals/components/step-route/test/index.tsx

Co-authored-by: daledupreez <dale@automattic.com>

gabrielcaires committed:
Remove extra comments

gabrielcaires committed:
Improve constant names on test file

daledupreez commented on client/landing/stepper/declarative-flow/site-migration-flow.ts:
```diff
@@ -41,51 +41,29 @@ const siteMigration: Flow = {
-			}
-		}, [ startUrl, userIsLoggedIn ] );
-
 		useEffect( () => {
 			if ( isOwner === false ) {
 				window.location.assign( '/start' );
 			}
 		}, [ isOwner ] );
```
I am seeing a race condition around this check for the flow. It feels more likely when I test using a URL like `/setup/site-migration/fr`.

I think we likely still need a check for the user's login state. It used to be implicit due to redirects before, but we should keep that in place.
```suggestion
		const userIsLoggedIn = useSelect(
			( select ) => ( select( USER_STORE ) as UserSelect ).isCurrentUserLoggedIn(),
			[]
		);

		useEffect( () => {
			if ( isOwner === false && userIsLoggedIn ) {
				window.location.assign( '/start' );
			}
		}, [ isOwner, userIsLoggedIn ] );
```

gabrielcaires replied:
Fixed here 7d080c58aea351810c3433d901f150e0649b4799

gabrielcaires committed:
Add missing type keyword

gabrielcaires committed:
Check user is logged in before redirect to start

daledupreez commented on client/landing/stepper/declarative-flow/site-migration-flow.ts:
```diff
@@ -41,51 +41,33 @@ const siteMigration: Flow = {
+		}, [ isOwner, userIsLoggedIn ] );
 
 		if ( ! siteSlug && ! siteId && ! isHostedSiteMigrationFlow( this.variantSlug ?? FLOW_NAME ) ) {
 			window.location.assign( '/start' );
-			result = {
+
+			return {
 				state: AssertConditionState.FAILURE,
 				message: 'site-setup did not provide the site slug or site id it is configured to.',
 			};
 		}
```
I am wondering if we have two issues in this code block:
1. This should be inside `useEffect()`
2. Should this code also be checking for whether we have a logged-in user?

daledupreez replied:
As a follow-up, I think this is another situation where we had an implicit dependency on the login redirect logic, which we've now removed.

daledupreez replied:
Fix pushed in 4614bf4de111ec61e5717f2782630594c2b32cfb.

daledupreez committed:
Update redirection for missing site details

gabrielcaires committed:
Merge branch 'trunk' into try/improve-login-redirects-and-step-tracking-in-stepper

gabrielcaires commented:
> It seems the /setup/site-migration flow is redirecting me to /start. Am I missing something?

Oh, let me check it.

gabrielcaires commented:
> It seems the /setup/site-migration flow is redirecting me to /start. Am I missing something?

You can't access this flow directly without having a site defined (siteId=[SiteId], siteSlug='someSlug')

daledupreez reviewed (commented):
@gabrielcaires, I have re-tested, and I think we are OK to get this shipped. I can't approve as the original PR author, but let me know if you have any concerns about shipping this as-is.
